### ISSUE TITLE
Define structure of an endpoint name

### ISSUE THREAD

**Issue by cabo** [core-wg/resource-directory#190]
All we know from the draft is that it is ≤ 63 bytes.  But we need to say that it is a UTF-8 string.  Not entirely sure we want 😈👄👽 as an endpoint name either.

Same for sector.

**Comment by cabo** [comment]
(via Jaime)

**Comment by chrysn** [comment]
We certainly want to allow people to use license plate numbers as endpoint names, so at very least we need to support [`ep=3MOJI😂` and `ep=B34UT😀`](https://www.businessinsider.de/australia-lets-people-use-emojis-on-car-license-plates-2019-2?r=US&IR=T) before they register as `ep=xn--3moji-g324d`.

More seriously, there are different uses of the endpoint name that may all warrant some restrictions to make them simpler (one could forbid the `"` (quotation mark) character to ease RFC6690 serialization, the `&` (ampersand) character to ease the representation of the registration operation's resource as a URI, forbid the `/` (solidus) to allow RDs to use the ep in the registration URI, forbid the `.` (full stop) character as DNS host names can contain dots but those can't be encoded in a URI), but at the same time there are applications for some of them (eg. dots are suggested in the NAI examples [of LwM2M](http://www.openmobilealliance.org/release/LightweightM2M/V1_1-20180710-A/HTML-Version/OMA-TS-LightweightM2M_Core-V1_1-20180710-A.html#7-3-1-0-731-Endpoint-Client-Name)).

Limitations to which Unicode code points are there would be hard to enforce reliably, and I don't see a compelling reason to just "rule out everything that's not ASCII". If anything, one could think about the code points 0-31 (except some combinations of whitespace) and 127, which can't be represented in RFC6690 as they aren't part of RFC2616's TEXT (whoever then misses `ep=THIS-IS-␛[31;1;4mRED␛[0m` ANSI-colored endpoint names can fall back to QuakeIII style of `ep=THIS-IS-^1RED^0` – those examples may sound mocking, but serve to illustrate that we won't easily catch all the characters that are bad anywhere without becoming very restrictive).

Being valid UTF8 is a requirement that we should just put in there (and check for others like sector). Any further limitations should IMO be policy-based, typically by the ep being limited to devices with known names anyway.

**Comment by petervanderstok** [comment]
UTF-8 or even more restricted. Certainly not impose a structure.....

**Comment by chrysn** [comment]
A suggestion from Wednesday's interim was to look into the URI/IRI ABNF rules for reasonable sets of characters.

The sets in RFCs [3986](https://tools.ietf.org/html/rfc3986)/[3987](https://tools.ietf.org/html/rfc3987) are not directly helpful as their production rules either rule out the LwM2M patterns of using URNs in there (which require at least ":" to be permitted) or bring in URI structure which would be imposing structure – the "[i]query" rule, for example, would only permit many URNs, but contains the pct-encoded rule that says percent signs are only OK if followed by two hex digits.

I've also looked into [RFC6570](https://tools.ietf.org/html/rfc6570) any restrictions we currently implicitly have would stem from there (given ep and d are passed in via URI template variable), but that's very relaxed, talks of strings and percent-encoding almost anything ("eplaced by its value as a string after pct-encoding any characters not in the set of unreserved URI characters").

For lack of better concrete options I'd reach into what can be used in link-format, phrased in RFC8288/7230 terms as HTAB / SP / 0x21-7E / 0x80-FF as long as it is valid UTF-8 (which is the innards of quoted-string with no exceptions for '"' or '\' as those are escaping issues) and prepare text for that.

(It also dawns to me that the cause of us having the problem of defining identifiers is that we did not think this with enough web in mind, that endpoint names should always be URIs (as that's the uniform identifier we have in the web) and that consequently a registration's endpoint name should not so much be an attribute but a typed link to the abstract definition of that endpoint – but given there are RDs out there already and we'd like to finish this draft, that can of worms stays closed for now.)

**Comment by cabo** [comment]
HT is evil.  Must not be used.  To me it now seems all of UTF-8 minus C0 (0x00..0x1f) minus 0x7F is desired.
```
epchar = %x20-7e / %x80-10FFFF
```
Or, if we also want to exclude C1, i.e. all 65 Unicode control characters:
```
epchar = %x20-7e / %xA0-10FFFF
```

**Comment by chrysn** [comment]
Closed by d07dde44d4875091dcd8c7eb0380e3675e339d7a which should have said "Closes" rather than "Contributes-To" in retrospect.
